# PHASE 3 PROJECT
## Tanzanian WaterWells
HOSTED BY DRIVENDATA

In [1]:
# Objective
# Business problem
# Business understanding

In [1]:
# Importing relevant libraries
import pandas as pd
pd.set_option("display.max_columns", None)
import numpy as np

In [2]:
# Importing the datasets
df1 = pd.read_csv("datasets/TZ-wells/training_set_values.csv")
df2 = pd.read_csv("datasets/TZ-wells/training_set_labels.csv")

### Explore
Checking the structure of the dataframes below. We start with the df1 dataframe:

In [13]:
print(df1.info())
df1.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 40 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   amount_tsh             59400 non-null  float64
 2   date_recorded          59400 non-null  object 
 3   funder                 55765 non-null  object 
 4   gps_height             59400 non-null  int64  
 5   installer              55745 non-null  object 
 6   longitude              59400 non-null  float64
 7   latitude               59400 non-null  float64
 8   wpt_name               59400 non-null  object 
 9   num_private            59400 non-null  int64  
 10  basin                  59400 non-null  object 
 11  subvillage             59029 non-null  object 
 12  region                 59400 non-null  object 
 13  region_code            59400 non-null  int64  
 14  district_code          59400 non-null  int64  
 15  lg

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  \
0  34.938093  -9.856322                  none            0   
1  34.698766  -2.147466              Zahanati            0   
2  37.460664  -3.821329           Kwa Mahundi            0   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
4  31.130847  -1.825359               Shuleni            0   

                     basin  subvillage   region  region_code  district_code  \
0               Lake Nyasa    Mnyusi B   Iringa           11              5   
1            Lake Victoria     Nyamara     Mara           20              2   
2                  Pangani     Majengo  Manyara           21              4   
3  Ruvuma / Southern Coast  Mahakamani   Mtwara           90             63   
4            Lake Victoria  Kyanyamisa   Kagera           18              1   

         lga        ward  population public_meeting              recorded_by  \
0     Ludewa    Mundindi         109           True  GeoData Consultants Ltd   
1  Serengeti       Natta         280            NaN  GeoData Consultants Ltd   
2  Simanjiro     Ngorika         250           True  GeoData Consultants Ltd   
3   Nanyumbu    Nanyumbu          58           True  GeoData Consultants Ltd   
4    Karagwe  Nyakasimbi           0           True  GeoData Consultants Ltd   

  scheme_management                  scheme_name permit  construction_year  \
0               VWC                        Roman  False               1999   
1             Other                          NaN   True               2010   
2               VWC  Nyumba ya mungu pipe scheme   True               2009   
3               VWC                          NaN   True               1986   
4               NaN                          NaN   True                  0   

  extraction_type extraction_type_group extraction_type_class management  \
0         gravity               gravity               gravity        vwc   
1         gravity               gravity               gravity        wug   
2         gravity               gravity               gravity        vwc   
3     submersible           submersible           submersible        vwc   
4         gravity               gravity               gravity      other   

  management_group         payment payment_type water_quality quality_group  \
0       user-group    pay annually     annually          soft          good   
1       user-group       never pay    never pay          soft          good   
2       user-group  pay per bucket   per bucket          soft          good   
3       user-group       never pay    never pay          soft          good   
4            other       never pay    never pay          soft          good   

       quantity quantity_group                source           source_type  \
0        enough         enough                spring                spring   
1  insufficient   insufficient  rainwater harvesting  rainwater harvesting   
2        enough         enough                   dam                   dam   
3           dry            dry           machine dbh              borehole   
4      seasonal       seasonal  rainwater harvesting  rainwater harvesting   

  source_class              waterpoint_type waterpoint_type_group  
0  groundwater           communal standpipe    communal standpipe  
1      surface           communal standpipe    communal standpipe  
2      surface  communal standpipe multiple    communal standpipe  
3  groundwater  communal standpipe multiple    communal standpipe 

In [11]:
# Checking the structure of df2
print(df2.info())
df2.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            59400 non-null  int64 
 1   status_group  59400 non-null  object
dtypes: int64(1), object(1)
memory usage: 928.2+ KB
None


id    status_group
0  69572      functional
1   8776      functional
2  34310      functional
3  67743  non functional
4  19728      functional

We observe that the two data frames are connecte and our target variable is in df2. We will merge these two dataframes into a combined dataframe before pre-processing and then we will split it into train and test sets for modeling.

## 2. EXPLORATORY DATA ANALYSIS
Below we perform EDA, first we will start by merging the two dataframes

In [3]:
# Merging the two dataframes
data = pd.merge(df1, df2)
data.head()

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572      6000.0    2011-03-14         Roman        1390         Roman   
1   8776         0.0    2013-03-06       Grumeti        1399       GRUMETI   
2  34310        25.0    2013-02-25  Lottery Club         686  World vision   
3  67743         0.0    2013-01-28        Unicef         263        UNICEF   
4  19728         0.0    2011-07-13   Action In A           0       Artisan   

   longitude   latitude              wpt_name  num_private  \
0  34.938093  -9.856322                  none            0   
1  34.698766  -2.147466              Zahanati            0   
2  37.460664  -3.821329           Kwa Mahundi            0   
3  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
4  31.130847  -1.825359               Shuleni            0   

                     basin  subvillage   region  region_code  district_code  \
0               Lake Nyasa    Mnyusi B   Iringa           11              5   
1            Lake Victoria     Nyamara     Mara           20              2   
2                  Pangani     Majengo  Manyara           21              4   
3  Ruvuma / Southern Coast  Mahakamani   Mtwara           90             63   
4            Lake Victoria  Kyanyamisa   Kagera           18              1   

         lga        ward  population public_meeting              recorded_by  \
0     Ludewa    Mundindi         109           True  GeoData Consultants Ltd   
1  Serengeti       Natta         280            NaN  GeoData Consultants Ltd   
2  Simanjiro     Ngorika         250           True  GeoData Consultants Ltd   
3   Nanyumbu    Nanyumbu          58           True  GeoData Consultants Ltd   
4    Karagwe  Nyakasimbi           0           True  GeoData Consultants Ltd   

  scheme_management                  scheme_name permit  construction_year  \
0               VWC                        Roman  False               1999   
1             Other                          NaN   True               2010   
2               VWC  Nyumba ya mungu pipe scheme   True               2009   
3               VWC                          NaN   True               1986   
4               NaN                          NaN   True                  0   

  extraction_type extraction_type_group extraction_type_class management  \
0         gravity               gravity               gravity        vwc   
1         gravity               gravity               gravity        wug   
2         gravity               gravity               gravity        vwc   
3     submersible           submersible           submersible        vwc   
4         gravity               gravity               gravity      other   

  management_group         payment payment_type water_quality quality_group  \
0       user-group    pay annually     annually          soft          good   
1       user-group       never pay    never pay          soft          good   
2       user-group  pay per bucket   per bucket          soft          good   
3       user-group       never pay    never pay          soft          good   
4            other       never pay    never pay          soft          good   

       quantity quantity_group                source           source_type  \
0        enough         enough                spring                spring   
1  insufficient   insufficient  rainwater harvesting  rainwater harvesting   
2        enough         enough                   dam                   dam   
3           dry            dry           machine dbh              borehole   
4      seasonal       seasonal  rainwater harvesting  rainwater harvesting   

  source_class              waterpoint_type waterpoint_type_group  \
0  groundwater           communal standpipe    communal standpipe   
1      surface           communal standpipe    communal standpipe   
2      surface  communal standpipe multiple    communal standpipe   
3  groundwater  communal standpipe multiple    communal standp

In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 59400 entries, 0 to 59399
Data columns (total 41 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   amount_tsh             59400 non-null  float64
 2   date_recorded          59400 non-null  object 
 3   funder                 55765 non-null  object 
 4   gps_height             59400 non-null  int64  
 5   installer              55745 non-null  object 
 6   longitude              59400 non-null  float64
 7   latitude               59400 non-null  float64
 8   wpt_name               59400 non-null  object 
 9   num_private            59400 non-null  int64  
 10  basin                  59400 non-null  object 
 11  subvillage             59029 non-null  object 
 12  region                 59400 non-null  object 
 13  region_code            59400 non-null  int64  
 14  district_code          59400 non-null  int64  
 15  lg

We see that the data has a good mix of numeric and object data types.

In [18]:
# Checking for missing values percentage
(data.isna().sum() / len(data)) * 100

id                        0.000000
amount_tsh                0.000000
date_recorded             0.000000
funder                    6.119529
gps_height                0.000000
installer                 6.153199
longitude                 0.000000
latitude                  0.000000
wpt_name                  0.000000
num_private               0.000000
basin                     0.000000
subvillage                0.624579
region                    0.000000
region_code               0.000000
district_code             0.000000
lga                       0.000000
ward                      0.000000
population                0.000000
public_meeting            5.612795
recorded_by               0.000000
scheme_management         6.526936
scheme_name              47.417508
permit                    5.144781
construction_year         0.000000
extraction_type           0.000000
extraction_type_group     0.000000
extraction_type_class     0.000000
management                0.000000
management_group    

A number of columns have missing values, with the scheme_name having the highest percentage of missing values. We will pay close attention to these columns first before trying to visualize and explore possible relationships between the different columns against the target variable. 

#### 1. scheme_name column
With the scheme name, we have no definitive way of transforming or filling in the missing values, my opinion is to make it a candidate for dropping. In addition it does not seem to be usefull to the question we are trying to answer. However tht is yet to be decided. We will drop the column in the next cell.

In [4]:
data.drop("scheme_name", axis= 1, inplace= True)

#### 2. funder column

In [22]:
data.funder.nunique()

1897

In [39]:
data.funder.value_counts()

Government Of Tanzania    9084
Other                     3635
Danida                    3114
Hesawa                    2202
Rwssp                     1374
                          ... 
Rarymond Ekura               1
Justine Marwa                1
Municipal Council            1
Afdp                         1
Samlo                        1
Name: funder, Length: 1898, dtype: int64

In [28]:
data[data.funder.isna()]

id  amount_tsh date_recorded funder  gps_height installer  \
34     41583         0.0    2011-02-23    NaN         -41       NaN   
43     19282         0.0    2013-01-15    NaN        1642       NaN   
47     13620         0.0    2011-07-27    NaN           0       NaN   
65     51072         0.0    2013-02-09    NaN        1415       NaN   
71     17386         0.0    2011-03-31    NaN           0       NaN   
...      ...         ...           ...    ...         ...       ...   
59357  46563         0.0    2013-02-19    NaN        1635       NaN   
59366  55232         0.0    2013-02-02    NaN        1541       NaN   
59370  14796       200.0    2013-01-29    NaN        1154       NaN   
59376  34716         0.0    2013-02-03    NaN        1581       NaN   
59397  37057         0.0    2011-04-11    NaN           0       NaN   

       longitude  latitude             wpt_name  num_private            basin  \
34     39.812912 -7.889986  Msikitini Wa Ijumaa            0           Rufiji   
43     34.967789 -4.628921         Mvae Primary            0         Internal   
47     33.540607 -9.172905           Mahakamani            0       Lake Nyasa   
65     34.621598 -5.173136               Nyambi            0         Internal   
71     34.462228 -8.575780    Kwa Manyusi Mlilo            0           Rufiji   
...          ...       ...                  ...          ...              ...   
59357  34.971841 -5.098362              Shabani            0         Internal   
59366  34.765729 -5.027725              Joshoni            0         Internal   
59370  30.058731 -4.902633       Village Office            0  Lake Tanganyika   
59376  34.821039 -5.076258              Nasingo            0         Internal   
59397  34.017087 -8.750434              Mashine            0           Rufiji   

        subvillage   region  region_code  district_code            lga  \
34     Kilombero B    Pwani           60             63          Mafia   
43        Mwarufyu  Singida           13              2  Singida Rural   
47     Mpandapanda    Mbeya           12              4         Rungwe   
65          Mfumbu  Singida           13              2  Singida Rural   
71         Lembuka    Mbeya           12              7        Mbarali   
...            ...      ...          ...            ...            ...   
59357   Mampando A  Singida           13              2  Singida Rural   
59366    Mungumaho  Singida           13              2  Singida Rural   
59370     Mazungwe   Kigoma           16              2         Kasulu   
59376           Hu  Singida           13              2  Singida Rural   
59397    Madungulu    Mbeya           12              7        Mbarali   

            ward  population public_meeting              recorded_by  \
34        Baleni         100           True  GeoData Consultants Ltd   
43         Merya           1           True  GeoData Consultants Ltd   
47        Kiwira           0           True  GeoData Consultants Ltd   
65      Naintiri           1           True  GeoData Consultants Ltd   
71       Mawindi           0           True  GeoData Consultants Ltd   
...          ...         ...            ...                      ...   
59357     Ntuntu           1           True  GeoData Consultants Ltd   
59366       Puma         200           True  GeoData Consultants Ltd   
59370     Rusesa           1           True  GeoData Consultants Ltd   
59376  Dung'unyi           1           True  GeoData Consultants Ltd   
59397    Chimala           0           True  GeoData Consultants Ltd   

      scheme_management scheme_name permit  construction_year extraction_type  \
34                  VWC         NaN  False                  0     nira/tanira   
43                  VWC           K    NaN               1980            mono   
47                  VWC           K    NaN                  0         gravity   
65                  VWC           M    NaN               1970            mono   
71                  VWC  

We can see from the cells above, the funder column contains 3635 values. It is a categorical column, I see it fit to impute another category "Other" which will indicate that the well was funded through a different initiative, this could be individual and private endeavors or Community initiated actions.

In [5]:
data.funder.fillna("Others", inplace= True)

In [66]:
data.funder.isna().sum()

0

In [59]:
data.funder.value_counts().to_frame().T

Government Of Tanzania  Others  Danida  Hesawa  Rwssp  World Bank  \
funder                    9084    3637    3114    2202   1374        1349   

        Kkkt  World Vision  Unicef  Tasaf  District Council  Dhv  \
funder  1287          1246    1057    877               843  829   

        Private Individual  Dwsp    0  Norad  Germany Republi  Tcrs  \
funder                 826   811  777    765              610   602   

        Ministry Of Water  Water  Dwe  Netherlands  Hifab  Adb  Lga  Amref  \
funder                590    583  484          470    450  448  442    425   

        Fini Water  Oxfam  Wateraid  Rc Church  Isf  Rudep  Mission  Private  \
funder         393    359       333        321  316    312      301      295   

        Jaica  Roman  Rural Water Supply And Sanitat  Adra  Ces(gmbh)  Jica  \
funder    280    275                             270   263        260   259   

        Shipo  Wsdp   Rc  Finw   Dh  Ded  Plan Int  Kiliwater  Dmdd   Go  \
funder    241   234  230   219  213  198       195        189   186  181   

        Lawatefuka Water Supply  Oxfarm  Magadini-makiwaru Water   Fw  W.B  \
funder                      180     176                      175  173  170   

        Kkkt_makwale  Ces (gmbh)  Wvt  Oikos E.Afrika  Nethalan  \
funder           166         165  158             155       152   

        Mkinga Distric Coun  Lvia  Concern World Wide  Unhcr   No  Swedish  \
funder                  150   147                 145    137  134      131   

        African  Anglican Church  Community   Is   He   Ki   Ir  Tardo  \
funder      128              126        126  125  125  124  123    123   

        Wananchi  Snv  Roman Catholic  Wua  Unice  Bsf  Tassaf   Co  Dfid  \
funder       122  122             121  121    119  117     116  111   110   

        Lamp  Concern  Muwsa  Villagers  Village Council   Ru  \
funder   110      108    107        106              105  105   

        Halmashauri Ya Wilaya Sikonge  Germany  Hsw  Twe  Idc  Tanza  Missi  \
funder                            102      101  101   97   92     88     87   

        Mdrdp   H  Undp  Aict  Gtz  Japan  Cmsr  Rc Ch  Vwc  Ndrdp  Kuwait  \
funder     86  86    86    85   84     82    81     80   78     78      72   

        Lwi  Fin Water  Caritas  Kkkt Church  Cdtf  Padep  Kaemp  Conce  \
funder   72         72       70           70    70     69     69     68   

        Water Aid /sema  Marafip  Kibaha Town Council  Cefa  Ncaa  Commu  \
funder               68       68                   68    67    63     62   

        Losaa-kia Water Supply  Mkinga  Distric Cou  National Rural  Md  \
funder                      62                   62              62  61   

        Irish Ai  Sabemo  Plan International  Twesa  Tlc  Gen  Idara Ya Maji  \
funder        59      59                  57     56   54   54             54   

        St  Kirde  Grumeti  Finida German Tanzania Govt  Solidarm  \
funder  54     54       53                           53        53   

        Kilindi District Co  Wfp  Cocen  China Government  Tanapa  Ta  \
funder                   53   52     52                51      51  51   

        Dwe/norad  Sema  Tabora Municipal Council  Solidame  European Union  \
funder         50    50                        49        48              48   

        Dwssp  Ridep  Acra  Dads  Devon Aid Korogwe  Miziriol  Shawasa  \
funder     47     47    46    46                 46        46       45   

        Red Cross  Redep  Tz Japan  Abasia  Giz  Ms  Donor  Cafod  Tdft  \
funder         45     44        43      43   42  42     41     41    41   

        Serikali  Soda  Un  Tredep  Ka  Songea District Council  Kuwasa  Jbg  \
funder        41    40  40      40  39                       38      38   38   

        Fathe  Dasip  Kidp   W  Urt  Islamic Found  Songea Municipal Counci  \
funder     37     37    36  36   36             35                       35   

        Mbiuwasa  African Development Bank  Watu Wa Ujerumani 

I have observed that there is a category which contains zero as its value class, this we will change to the Unknown category. Also there are some string values which could be standardized to ensure that the values are uniform. For example, we could change the values of Roman Cathoric, Roman Church to Catholic, Angrikana to Anglican, Rwsssp to Rwssp, Serikari to Government Of Tanzania

In [60]:
data[data.funder == '0']

id  amount_tsh date_recorded funder  gps_height installer  \
145    28669        50.0    2013-03-21      0          12         0   
152    60983         0.0    2013-03-16      0         -15         0   
393    39749         0.0    2013-03-18      0          28         0   
417    15832        50.0    2013-03-22      0          30         0   
428    50233         0.0    2013-03-12      0          30         0   
...      ...         ...           ...    ...         ...       ...   
59237   2138        59.0    2013-03-19      0          81         0   
59243   3396        50.0    2013-03-16      0         -20         0   
59276  62818        50.0    2013-03-21      0          18         0   
59351  55322        50.0    2013-03-18      0         -19         0   
59387  26640       100.0    2013-03-12      0          25         0   

       longitude  latitude       wpt_name  num_private        basin  \
145    39.185710 -6.892593       Chekanao            0  Wami / Ruvu   
152    39.527114 -6.988748      Msikitini            0  Wami / Ruvu   
393    39.159887 -6.902548   Kwa Chambuso            0  Wami / Ruvu   
417    39.178404 -6.938013     Ccm Kivule            0  Wami / Ruvu   
428    39.178849 -6.973206  Ofisi Ya Kata            0  Wami / Ruvu   
...          ...       ...            ...          ...          ...   
59237  39.119109 -6.898919           Kata            0  Wami / Ruvu   
59243  39.524021 -6.984802   Kwa Mariwala            0  Wami / Ruvu   
59276  39.183790 -6.897566   Kwa Mkunduge            0  Wami / Ruvu   
59351  39.534599 -7.088183      Kwa China            0  Wami / Ruvu   
59387  39.176480 -6.957098     Kwa Maliba            0  Wami / Ruvu   

                 subvillage         region  region_code  district_code  \
145    Mtaa Wa Kitunda Kati  Dar es Salaam            7              2   
152                 Madrasa  Dar es Salaam            7              3   
393        Mtaa Wa Mji Mpya  Dar es Salaam            7              2   
417          Mtaa Wa Kivule  Dar es Salaam            7              2   
428         Mtaa Yangeyange  Dar es Salaam            7              2   
...                     ...            ...          ...            ...   
59237       Mtaa Wa Bombani  Dar es Salaam            7              2   
59243          Miembe Dekwa  Dar es Salaam            7              3   
59276  Mtaa Wa Kitunda Kati  Dar es Salaam            7              2   
59351               Bondeni  Dar es Salaam            7              3   
59387    Mtaa Wa Yangeyange  Dar es Salaam            7              2   

          lga         ward  population public_meeting  \
145     Ilala      Kitunda         150            NaN   
152    Temeke      Kimbiji          50           True   
393     Ilala      Kitunda           1            NaN   
417     Ilala     Msongola         100            NaN   
428     Ilala     Msongola        1000           True   
...       ...          ...         ...            ...   
59237   Ilala         Pugu         150            NaN   
59243  Temeke      Kimbiji         210           True   
59276   Ilala      Kitunda         250            NaN   
59351  Temeke  Pemba Mnazi          50           True   
59387   Ilala     Msongola         100            NaN   

                   recorded_by scheme_management permit  construction_year  \
145    GeoData Consultants Ltd  Private operator  False                  0   
152    GeoData Consultants Ltd               VWC  False               1970   
393    GeoData Consultants Ltd               NaN  False               2010   
417    GeoData Consultants Ltd               WUG  False               2000   
428    GeoData Consultants Ltd             Other  False               2000   
...                        ...               ...    ...                ...   
59237  GeoData Consultants Ltd               WUG  False               2000   
59243  GeoData Consultants Ltd               VWC  False               2010   
59276  GeoData Consultants L

In [49]:
data[data.funder == 'None']

id  amount_tsh date_recorded funder  gps_height installer  \
4868   52471         0.0    2012-10-13   None           0       NaN   
19055  18348         0.0    2012-10-14   None           0       NaN   

       longitude  latitude   wpt_name  num_private     basin   subvillage  \
4868   33.357464 -4.022400  Kwa Jacob            0  Internal   Uswahilini   
19055  33.459053 -4.170978  Kwa Abeli            0  Internal  Mwandumpolo   

       region  region_code  district_code     lga   ward  population  \
4868   Tabora           14              2  Igunga  Choma           0   
19055  Tabora           14              2  Igunga  Ngulu           0   

      public_meeting              recorded_by scheme_management scheme_name  \
4868            True  GeoData Consultants Ltd   Water authority  World Bank   
19055           True  GeoData Consultants Ltd             Other         NaN   

      permit  construction_year extraction_type extraction_type_group  \
4868    True                  0     submersible           submersible   
19055  False                  0          swn 80                swn 80   

      extraction_type_class management management_group         payment  \
4868            submersible        vwc       user-group  pay per bucket   
19055              handpump        vwc       user-group         unknown   

      payment_type water_quality quality_group      quantity quantity_group  \
4868    per bucket         salty         salty        enough         enough   
19055      unknown         salty         salty  insufficient   insufficient   

             source   source_type source_class     waterpoint_type  \
4868    machine dbh      borehole  groundwater  communal standpipe   
19055  shallow well  shallow well  groundwater           hand pump   

      waterpoint_type_group    status_group  
4868     communal standpipe      functional  
19055             hand pump  non functional

In [ ]:
# Define a dictionary below
funders_dict = {
    ""
}

#### 3. Installer
This column can be used to explain if the wells were installed through specialized help or otherwise. However there is no way to confirm if the group involved are certified professionals or not. It could be usefull if we are trying to find the functionality of wells in regards to who installed it and if they were qualified or not.

In [75]:
data.installer.nunique()

2145

In [76]:
data.installer.value_counts().to_frame().T

DWE  Government   RWE  Commu  DANIDA  KKKT  Hesawa    0  TCRS  \
installer  17402        1825  1206   1060    1050   898     840  777   707   

           Central government  CES  Community  DANID  District Council  \
installer                 622  610        553    552               551   

           HESAWA  LGA  World vision  WEDECO  TASAF  District council  Gover  \
installer     539  408           408     397    396               392    383   

           AMREF  TWESA   WU  Dmdd  ACRA  World Vision  SEMA   DW  OXFAM   Da  \
installer    329    316  301   287   278           270   249  246    234  224   

           Gove  UNICEF  Idara ya maji  Sengerema Water Department  Kiliwater  \
installer   222     222            222                         214        210   

           NORAD  FinW   DH  Villagers  DWSP  Distri  Lawatefuka water sup  \
installer    208   208  202        199   192     181                   180   

           Magadini-Makiwaru wa   RC   FW  KKKT _ Konde and DWE  Centr  WVT  \
installer                   175  174  173                   166    162  158   

           MWE  Handeni Trunk Main(   Is  Norad  RWSSP  Fini Water  SHIPO  \
installer  157                  156  154    152    149         149    147   

           Private  Kuwait  JAICA  Central govt  ISF  Artisan  Fini water  \
installer      143     142    141           138  135      135         133   

           GOVER  DDCA  RC CHURCH   HE  Tardo  World  Mission  wananchi   Ir  \
installer    128   126        125  125    123    121      121       119  119   

           Consulting Engineer  Amref  JICA  TWE  Central Government  MUWSA  \
installer                  116    114   113  111                 110    107   

           DED  WATER AID  FINI WATER  Halmashauri ya wilaya sikonge  \
installer  105        103         103                            102   

           Wizara ya maji  HSW  Jaica  Unisef  Ki  Go  OXFARM  World Bank  \
installer             100  100     98      98  97  97      96          95   

           Roman  Angli  VWC  District water department  DMDD  Missi   H  \
installer     94     92   91                         89    89     87  86   

           Shipo  AICT  DA  MDRDP  Water board  Adra  Oikos E .Africa  \
installer     86    84  84     84           81    80               80   

           District Water Department  Region water Department  Chamavita  \
installer                         78                       77         75   

           RC Church  CEFA  Danid  Water  LWI  Wa  KKKT CHURCH  GOVERNMENT  \
installer         75    75     72     71   71  69           69          66   

           Villa  MTN  TASSAF  DAWASCO  GTZ  Losaa-Kia water supp  Finwater  \
installer     65   65      64       64   63                    62        62   

           Wachina  Adra/Community  LVIA  FIN WATER  Village Council  NCAA  \
installer       60              60    60         59               58    57   

           Adra /Community  GEN  TLC  Co  RDC  ADRA  DHV  St  Priva  Communit  \
installer               55   55   55  55   55    55   54  54     54        54   

           Pr  wanan  Local te  RC Ch  RWE/DWE  CDTF  Water Aid  \
installer  53     52        52     52       52    50         48   

           Local  technician  KUWAIT  WORLD BANK  MLADE  Miziriol  \
installer                 47      47          47     46        46   

           District water depar  Plan Internationa  KA  SOLIDAME  Local  \
installer                    46                 46  46        45     45   

           SHAWASA  Water Aid /sema  TDFT  VILLAGE COUNCIL  Ga  \
installer       45               44    42               42  42   

           JANDU PLUMBER CO  CBHCC  MWE &  GRUMETI  KUWASA  KILI WATER  JBG  \
installer                41     41     40       39      38          38   37   

           SHY BUILDERS  RED CROSS  KIDP  MBIUWASA  BSF  HAPA SINGIDA  Ch  \
installer            37         36    36        35   35            34  34   

           FI

In [ ]:
# standardaze the column entries
# fix a few names

#### 4. date_recorded & construction_year

In [77]:
data.date_recorded.head()

0    2011-03-14
1    2013-03-06
2    2013-02-25
3    2013-01-28
4    2011-07-13
Name: date_recorded, dtype: object

From above the column is of the object type. The column contains dates on which the records were being inserted in the dataset. The column again does not seem to be useful on what we are trying to predict. I would recommend dropping it. But first we could use it to obtain the age of the wells at the time of recording information on particular wells. To do that we would need to obtain the year and then convert it to the int data type.

In [6]:
# Slicing the Series then converting it to int type
data.date_recorded = [x[:4] for x in data.date_recorded]
data.date_recorded = data.date_recorded.astype(str).astype(float).astype(int)
data.date_recorded.info()

<class 'pandas.core.series.Series'>
Int64Index: 59400 entries, 0 to 59399
Series name: date_recorded
Non-Null Count  Dtype
--------------  -----
59400 non-null  int64
dtypes: int64(1)
memory usage: 928.1 KB


#### construction_year
Now we work on the construction_year column. We will use these two columns and try to obtain the age of the well.

In [12]:
data.construction_year.info()

<class 'pandas.core.series.Series'>
Int64Index: 59400 entries, 0 to 59399
Series name: construction_year
Non-Null Count  Dtype
--------------  -----
59400 non-null  int64
dtypes: int64(1)
memory usage: 928.1 KB


In [81]:
data.construction_year.nunique()

55

In [11]:
data.construction_year.value_counts().sort_index()

0       20709
1960      102
1961       21
1962       30
1963       85
1964       40
1965       19
1966       17
1967       88
1968       77
1969       59
1970      411
1971      145
1972      708
1973      184
1974      676
1975      437
1976      414
1977      202
1978     1037
1979      192
1980      811
1981      238
1982      744
1983      488
1984      779
1985      945
1986      434
1987      302
1988      521
1989      316
1990      954
1991      324
1992      640
1993      608
1994      738
1995     1014
1996      811
1997      644
1998      966
1999      979
2000     2091
2001      540
2002     1075
2003     1286
2004     1123
2005     1011
2006     1471
2007     1587
2008     2613
2009     2533
2010     2645
2011     1256
2012     1084
2013      176
Name: construction_year, dtype: int64

In [89]:
(data.construction_year.value_counts()[0] / len(data)) * 100

34.86363636363636

We have about 35 percent of the data being withought a known date of construction of the well. We will fill this values with the median year which is 1985.

In [13]:
data.loc[data.construction_year == 0, "construction_year"] = 1985
data["well_age"] = (data.date_recorded - data.construction_year)

In [19]:
# Dropping the two columns used above
data.drop(columns= ["date_recorded", "construction_year"], axis= 1, inplace= True)

In [20]:
data.head()

id  amount_tsh        funder  gps_height     installer  longitude  \
0  69572      6000.0         Roman        1390         Roman  34.938093   
1   8776         0.0       Grumeti        1399       GRUMETI  34.698766   
2  34310        25.0  Lottery Club         686  World vision  37.460664   
3  67743         0.0        Unicef         263        UNICEF  38.486161   
4  19728         0.0   Action In A           0       Artisan  31.130847   

    latitude              wpt_name  num_private                    basin  \
0  -9.856322                  none            0               Lake Nyasa   
1  -2.147466              Zahanati            0            Lake Victoria   
2  -3.821329           Kwa Mahundi            0                  Pangani   
3 -11.155298  Zahanati Ya Nanyumbu            0  Ruvuma / Southern Coast   
4  -1.825359               Shuleni            0            Lake Victoria   

   subvillage   region  region_code  district_code        lga        ward  \
0    Mnyusi B   Iringa           11              5     Ludewa    Mundindi   
1     Nyamara     Mara           20              2  Serengeti       Natta   
2     Majengo  Manyara           21              4  Simanjiro     Ngorika   
3  Mahakamani   Mtwara           90             63   Nanyumbu    Nanyumbu   
4  Kyanyamisa   Kagera           18              1    Karagwe  Nyakasimbi   

   population public_meeting              recorded_by scheme_management  \
0         109           True  GeoData Consultants Ltd               VWC   
1         280            NaN  GeoData Consultants Ltd             Other   
2         250           True  GeoData Consultants Ltd               VWC   
3          58           True  GeoData Consultants Ltd               VWC   
4           0           True  GeoData Consultants Ltd               NaN   

  permit extraction_type extraction_type_group extraction_type_class  \
0  False         gravity               gravity               gravity   
1   True         gravity               gravity               gravity   
2   True         gravity               gravity               gravity   
3   True     submersible           submersible           submersible   
4   True         gravity               gravity               gravity   

  management management_group         payment payment_type water_quality  \
0        vwc       user-group    pay annually     annually          soft   
1        wug       user-group       never pay    never pay          soft   
2        vwc       user-group  pay per bucket   per bucket          soft   
3        vwc       user-group       never pay    never pay          soft   
4      other            other       never pay    never pay          soft   

  quality_group      quantity quantity_group                source  \
0          good        enough         enough                spring   
1          good  insufficient   insufficient  rainwater harvesting   
2          good        enough         enough                   dam   
3          good           dry            dry           machine dbh   
4          good      seasonal       seasonal  rainwater harvesting   

            source_type source_class              waterpoint_type  \
0                spring  groundwater           communal standpipe   
1  rainwater harvesting      surface           communal standpipe   
2                   dam      surface  communal standpipe multiple   
3              borehole  groundwater  communal standpipe multiple   
4  rainwater harvesting      surface           communal standpipe   

  waterpoint_type_group    status_group  well_age  
0    communal standpipe      functional        12  
1    communal standpipe      functional         3  
2    communal standpipe      functional         4  
3    communal standpipe  non functional        27  
4    communal standpipe      functional        26

#### 5. gps_height
This column contains the altitude of the well, which is the height of the well above sea level.

In [21]:
data.gps_height.nunique()

2428

In [24]:
data.gps_height.value_counts().sort_index().to_frame().T

-90    -63    -59    -57    -55    -54    -53    -52    -51    \
gps_height      1      2      1      1      1      1      1      2      2   

            -50    -49    -48    -47    -46    -45    -44    -43    -42    \
gps_height      5      4      6      8      6      9      6     11     13   

            -41    -40    -39    -38    -37    -36    -35    -34    -33    \
gps_height     11     26     14     17     19     24     18     18     22   

            -32    -31    -30    -29    -28    -27    -26    -25    -24    \
gps_height     24     26     26     26     36     42     24     34     33   

            -23    -22    -21    -20    -19    -18    -17    -16    -15    \
gps_height     45     38     40     52     47     49     44     55     60   

            -14    -13    -12    -11    -10    -9     -8     -7     -6     \
gps_height     51     55     36     41     28     38     44     38     41   

            -5     -4     -3     -2     -1      0      1      2      3     \
gps_height     38     36     38     40     22  20438     34     28     23   

             4      5      6      7      8      9      10     11     12    \
gps_height     24     33     33     33     24     25     26     26     41   

             13     14     15     16     17     18     19     20     21    \
gps_height     30     30     32     32     39     36     30     28     25   

             22     23     24     25     26     27     28     29     30    \
gps_height     22     34     19     30     35     27     27     21     34   

             31     32     33     34     35     36     37     38     39    \
gps_height     16     19     23     20     24     28     21     25     23   

             40     41     42     43     44     45     46     47     48    \
gps_height     25     21     20     26     28     28     21     19     31   

             49     50     51     52     53     54     55     56     57    \
gps_height     27     24     24     28     31     32     24     28     25   

             58     59     60     61     62     63     64     65     66    \
gps_height     22     24     30     21     30     19     23     17     23   

             67     68     69     70     71     72     73     74     75    \
gps_height     24     29     15     18     19     17     24     16     18   

             76     77     78     79     80     81     82     83     84    \
gps_height     21     15     19     14     26     17     23     19     16   

             85     86     87     88     89     90     91     92     93    \
gps_height     24     20     21     21     14     21     17     19     18   

             94     95     96     97     98     99     100    101    102   \
gps_height     12     14     23      7     18     18     19     16     24   

             103    104    105    106    107    108    109    110    111   \
gps_height     19     15     19     15     13     12     12     15      9   

             112    113    114    115    116    117    118    119    120   \
gps_height     14      6     14     19     17     16     20     12     11   

             121    122    123    124    125    126    127    128    129   \
gps_height     15     11     11     13     16     16      8     11      5   

             130    131    132    133    134    135    136    137    138   \
gps_height      8     13     12      9     12     12      9      6      6   

             139    140    141    142    143    144    145    146    147   \
gps_height      7     19      9     11     10     14      7      9      5   

             148    149    150    151    152    153    154    155    156   \
gps_height      7     11     11     15      9     11      8     14     17   

             157    158    159    160    161    162    163    164    165   \
gps_height     10     12     11      8     14      9     12     14     11   

             166    167    168    169    170    171    172    173    174   \
gps_height     10     12     11     14     13     12     21 

I don't think there is much we can do for this column since both the positive and negative values are in indication of the whether the well is below or above sea level. We could create categories from the column by coming up with a column which indicates if a well is at sea level, below or above sea level. In this next cell I define a function to do just that:

In [25]:
def sea_level(altitude):
    """This function creates categories for the columns
    then assigns the level above sea level.
    """
    if altitude < 0:
        return "below_sealevel"
    elif altitude == 0:
        return "at_sealevel"
    elif 0 < altitude <= 500:
        return "below_500m_sealevel"
    elif 500 < altitude <= 1500:
        return "below_1500m_sealevel"
    elif 1500 < altitude <= 2500:
        return "below_2500m_sealevel"
    else:
        return "above_2500m_sealevel"

In [26]:
data["sealevel"] = data["gps_height"].apply(sea_level)

In [27]:
data.head()

id  amount_tsh        funder  gps_height     installer  longitude  \
0  69572      6000.0         Roman        1390         Roman  34.938093   
1   8776         0.0       Grumeti        1399       GRUMETI  34.698766   
2  34310        25.0  Lottery Club         686  World vision  37.460664   
3  67743         0.0        Unicef         263        UNICEF  38.486161   
4  19728         0.0   Action In A           0       Artisan  31.130847   

    latitude              wpt_name  num_private                    basin  \
0  -9.856322                  none            0               Lake Nyasa   
1  -2.147466              Zahanati            0            Lake Victoria   
2  -3.821329           Kwa Mahundi            0                  Pangani   
3 -11.155298  Zahanati Ya Nanyumbu            0  Ruvuma / Southern Coast   
4  -1.825359               Shuleni            0            Lake Victoria   

   subvillage   region  region_code  district_code        lga        ward  \
0    Mnyusi B   Iringa           11              5     Ludewa    Mundindi   
1     Nyamara     Mara           20              2  Serengeti       Natta   
2     Majengo  Manyara           21              4  Simanjiro     Ngorika   
3  Mahakamani   Mtwara           90             63   Nanyumbu    Nanyumbu   
4  Kyanyamisa   Kagera           18              1    Karagwe  Nyakasimbi   

   population public_meeting              recorded_by scheme_management  \
0         109           True  GeoData Consultants Ltd               VWC   
1         280            NaN  GeoData Consultants Ltd             Other   
2         250           True  GeoData Consultants Ltd               VWC   
3          58           True  GeoData Consultants Ltd               VWC   
4           0           True  GeoData Consultants Ltd               NaN   

  permit extraction_type extraction_type_group extraction_type_class  \
0  False         gravity               gravity               gravity   
1   True         gravity               gravity               gravity   
2   True         gravity               gravity               gravity   
3   True     submersible           submersible           submersible   
4   True         gravity               gravity               gravity   

  management management_group         payment payment_type water_quality  \
0        vwc       user-group    pay annually     annually          soft   
1        wug       user-group       never pay    never pay          soft   
2        vwc       user-group  pay per bucket   per bucket          soft   
3        vwc       user-group       never pay    never pay          soft   
4      other            other       never pay    never pay          soft   

  quality_group      quantity quantity_group                source  \
0          good        enough         enough                spring   
1          good  insufficient   insufficient  rainwater harvesting   
2          good        enough         enough                   dam   
3          good           dry            dry           machine dbh   
4          good      seasonal       seasonal  rainwater harvesting   

            source_type source_class              waterpoint_type  \
0                spring  groundwater           communal standpipe   
1  rainwater harvesting      surface           communal standpipe   
2                   dam      surface  communal standpipe multiple   
3              borehole  groundwater  communal standpipe multiple   
4  rainwater harvesting      surface           communal standpipe   

  waterpoint_type_group    status_group  well_age              sealevel  
0    communal standpipe      functional        12  below_1500m_sealevel  
1    communal standpipe      functional         3  below_1500m_sealevel  
2    communal standpipe      functional         4  below_1500m_sealevel  
3    communal standpipe  non functional        27   below_500m_sealevel  
4    communal standpipe      functional        26           at_sealevel

In [56]:
(data.isna().sum() / len(data)) * 100

id                        0.000000
amount_tsh                0.000000
date_recorded             0.000000
funder                    0.000000
gps_height                0.000000
installer                 6.153199
longitude                 0.000000
latitude                  0.000000
wpt_name                  0.000000
num_private               0.000000
basin                     0.000000
subvillage                0.624579
region                    0.000000
region_code               0.000000
district_code             0.000000
lga                       0.000000
ward                      0.000000
population                0.000000
public_meeting            5.612795
recorded_by               0.000000
scheme_management         6.526936
scheme_name              47.417508
permit                    5.144781
construction_year         0.000000
extraction_type           0.000000
extraction_type_group     0.000000
extraction_type_class     0.000000
management                0.000000
management_group    